homeassistant/components/conversation/default_agent.py

```diff
@@ -118,25 +75,119 @@ async def async_process(
         language: str | None = None,
     ) -> ConversationResult | None:
         """Process a sentence."""
-        intents = self.hass.data[DOMAIN]
-
-        for intent_type, matchers in intents.items():
-            for matcher in matchers:
-                if not (match := matcher.match(text)):
-                    continue
-
-                intent_response = await intent.async_handle(
-                    self.hass,
-                    DOMAIN,
-                    intent_type,
-                    {key: {"value": value} for key, value in match.groupdict().items()},
-                    text,
-                    context,
-                    language,
-                )
-
-                return ConversationResult(
-                    response=intent_response, conversation_id=conversation_id
-                )
+        language = language or self.hass.config.language
+        lang_intents = await self.async_get_or_load_intents(language)
+        if lang_intents is None:
+            # No intents loaded
+            return None
+
+        slot_lists: dict[str, SlotList] = {
+            "area": self._make_areas_list(),
+            "name": self._make_names_list(),
+        }
+
+        result = recognize(text, lang_intents, slot_lists=slot_lists)
+        if result is not None:
+            intent_response = await intent.async_handle(
+                self.hass,
+                DOMAIN,
+                result.intent.name,
+                {
+                    entity.name: {"value": entity.value}
+                    for entity in result.entities_list
+                },
+                text,
+                context,
+                language,
+            )
+
+            return ConversationResult(
+                response=intent_response, conversation_id=conversation_id
+            )
 
         return None
+
+    async def async_get_or_load_intents(self, language: str) -> Intents | None:
+        """Load all intents for language."""
+        lang_intents = self._lang_intents.get(language)
+
+        if lang_intents is None:
+            # Load all sentences for language
```

Comment:
This code should be done in a single function in the executor. Load all integrations first in parallel using an async.gather